### ISSUE TITLE
No __cuda candidates found despite cuda being detected

### ISSUE THREAD

**Issue by msegado** [prefix-dev/pixi#607]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://github.com/prefix-dev/pixi/releases) of pixi, using `pixi --version`.

### Reproducible example

pixi.toml: _[edit: copy-pasted the wrong version in my original post]_
```toml
[project]
name = "bayes"
version = "0.1.0"
description = "Add a short description here"
channels = ["nvidia", "conda-forge"]
platforms = ["linux-64"]

[system-requirements]
cuda = "11.6"
linux = "3.10.0"

[tasks]

[dependencies]
```

Shell commands and error:
```shell
$ pixi info
pixi 0.10.0

Platform            : linux-64
Virtual packages    : __unix=0=0
                    : __linux=3.10.0=0
                    : __glibc=2.17=0
                    : __cuda=11.6=0
                    : __archspec=1=x86_64
Cache dir           : /<homedir>/.cache/rattler/cache
Auth storage        : /<homedir>/.rattler/rattler_auth_store.json

Project
------------

Manifest file       : /<workdir>/pixi.toml
Dependency count    : 0
Last updated        : 02-01-2024 14:22:28
Target platforms    : linux-64



$ pixi add jaxlib=*=*cuda*
  × could not determine any available versions for jaxlib on linux-64. Either the package could not be found or version constraints on other dependencies result in
  │ a conflict.
  ╰─▶ Cannot solve the request because of: jaxlib * *cuda* cannot be installed because there are no viable options:
      |-- jaxlib 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10 | 0.3.10
      | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14
      | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14
      | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.14 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15
      | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15
      | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.15 | 0.3.20 | 0.3.20 | 0.3.20 | 0.3.20 | 0.3.20
      | 0.3.20 | 0.3.20 | 0.3.20 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22
      | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.22 | 0.3.24 | 0.3.24 | 0.3.24 | 0.3.24 | 0.3.24 | 0.3.24 | 0.3.24 | 0.3.24 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 |
      0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.3.25 | 0.4.1 | 0.4.1 | 0.4.1 | 0.4.1 | 0.4.1 | 0.4.1 | 0.4.1 |
      0.4.1 | 0.4.2 | 0.4.2 | 0.4.2 | 0.4.2 | 0.4.2 | 0.4.2 | 0.4.2 | 0.4.2 | 0.4.3 | 0.4.3 | 0.4.3 | 0.4.3 | 0.4.3 | 0.4.3 | 0.4.3 | 0.4.4 | 0.4.4 | 0.4.4 | 0.4.6
      | 0.4.6 | 0.4.6 | 0.4.6 | 0.4.6 | 0.4.6 | 0.4.6 | 0.4.6 | 0.4.7 | 0.4.7 | 0.4.7 | 0.4.7 | 0.4.7 | 0.4.7 | 0.4.7 | 0.4.7 | 0.4.9 | 0.4.9 | 0.4.10 | 0.4.10
      | 0.4.10 | 0.4.10 | 0.4.11 | 0.4.11 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.12 | 0.4.14 | 0.4.14
      | 0.4.14 | 0.4.14 | 0.4.14 | 0.4.14 | 0.4.14 | 0.4.14 | 0.4.18 | 0.4.18 | 0.4.18 | 0.4.18 | 0.4.19 | 0.4.19 | 0.4.19 | 0.4.19 | 0.4.20 | 0.4.20 | 0.4.20 |
      0.4.23 | 0.4.23 | 0.4.23 would require
          |-- __cuda *, for which no candidates were found.
```

### Issue description

Pixi fails to install a cuda-enabled version of `jaxlib`, seemingly due to the resolver not finding a candidate for `__cuda`. However, cuda is indeed detected per `pixi info`, and also appears as a system requirement in `pixi.toml`.

Looks like the folks here are running into the same issue: https://github.com/ami-iit/jaxsim/pull/63#issuecomment-1870464739

Related: https://github.com/prefix-dev/pixi/issues/294

### Expected behavior

The package should be added and installed on a cuda-enabled system.

**Comment by baszalmstra** [comment]
The reason this happens is that even though your machine might support a specific version of cuda. Other machines might not. Pixi creates a lock-file that is guaranteed to work on all machine given certain minimum requirements. Not all machines support cuda so cuda is not part of those minimum requirements. However, you can modify these on a per project (or per platform) basis by adding the following to your project:

```toml
[system-requirements]
cuda = "11" # or any other version of cuda you want to use
```

See https://pixi.sh/configuration/#using-cuda-in-pixi

I'm going to close this issue, feel free to reopen if you feel this is still a bug.

**Comment by msegado** [comment]
Gah sorry, I copy-pasted the wrong version of my `pixi.toml` (will edit to update it now)... I did in fact have cuda = "11.6" in system-requirements. (I also tried without it, as well as with just cuda = "11"... all give the same error.)

**Comment by msegado** [comment]
@baszalmstra I can't reopen since I'm not a collaborator, but I think I found the bug. See line 412 of the `pixi add` command:

https://github.com/prefix-dev/pixi/blob/5a16570ae4af85b4de5b5fb858050e4c1e59b132/src/cli/add.rs#L403-L421

This seems to be using the default minimal set of virtual packages for solving (as returned by [`get_minimal_virtual_packages`](https://github.com/prefix-dev/pixi/blob/5a16570ae4af85b4de5b5fb858050e4c1e59b132/src/virtual_packages.rs#L28C8-L28C36)), rather than using the actual virtual packages for the project.

I think a fix would just be changing the `add` command to use [`Project::virtual_packages`](https://github.com/prefix-dev/pixi/blob/5a16570ae4af85b4de5b5fb858050e4c1e59b132/src/virtual_packages.rs#L110) instead, right?

**Comment by baszalmstra** [comment]
Ah yes I think you are right! Nice find! 😄 

Would you be able to make a PR?

**Comment by msegado** [comment]
> Ah yes I think you are right! Nice find! 😄

Thanks!

> Would you be able to make a PR?

I can definitely try 🙂  I'm not really a rustacean, but I made the change and the tests still pass so I'm hopeful...

**Comment by baszalmstra** [comment]
Fixed by #609

**Comment by traversaro** [comment]
> Looks like the folks here are running into the same issue: [ami-iit/jaxsim#63 (comment)](https://github.com/ami-iit/jaxsim/pull/63#issuecomment-1870464739)

Thanks for the cross-link, fyi @flferretti
